client/templates/typescript-fetch-templates/runtime.mustache

```diff
@@ -0,0 +1,344 @@
+            }
+            return `${encodeURIComponent(fullKey)}=${encodeURIComponent(String(value))}`;
+        })
+        .filter(part => part.length > 0)
+        .join('&');
+}
+
+export function mapValues(data: any, fn: (item: any) => any) {
+  return Object.keys(data).reduce(
+    (acc, key) => ({ ...acc, [key]: fn(data[key]) }),
+    {}
+  );
+}
+
+export function canConsumeForm(consumes: Consume[]): boolean {
+    for (const consume of consumes) {
+        if ('multipart/form-data' === consume.contentType) {
+            return true;
+        }
+    }
+    return false;
+}
+
+export interface Consume {
+    contentType: string
+}
+
+export interface RequestContext {
+    fetch: FetchAPI;
+    url: string;
+    init: RequestInit;
+}
+
+export interface ResponseContext {
+    fetch: FetchAPI;
+    url: string;
+    init: RequestInit;
+    response: Response;
+}
+
+export interface Middleware {
+    pre?(context: RequestContext): Promise<FetchParams | void>;
+    post?(context: ResponseContext): Promise<Response | void>;
+}
+
+export interface ApiResponse<T> {
+    raw: Response;
+    value(): Promise<T>;
+}
+
+export interface ResponseTransformer<T> {
+    (json: any): T;
+}
+
+export class JSONApiResponse<T> {
+    constructor(public raw: Response, private transformer: ResponseTransformer<T> = (jsonValue: any) => jsonValue) {}
+
+    async value(): Promise<T> {
+        return this.transformer(await this.raw.json());
+    }
+}
+
+export class VoidApiResponse {
+    constructor(public raw: Response) {}
+
+    async value(): Promise<void> {
+        return undefined;
+    }
+}
+
+export class BlobApiResponse {
+    constructor(public raw: Response) {}
+
+    async value(): Promise<Blob> {
+        return await this.raw.blob();
+    };
+}
+
+export class TextApiResponse {
+    constructor(public raw: Response) {}
+
+    async value(): Promise<string> {
+        return await this.raw.text();
+    };
+}
+/** CUSTOM STARTS HERE! **/
+
+/**
+ * When the rest client raises an error, the Error's message would be a json string of this format:
+ */
+export class RestClientCallError extends Error {
+    constructor(message: string, public readonly statusCode: number, public readonly statusMessage: string, public readonly body: string) {
+        super(message);
+    }
+}
+
+export class RestClientUtils {
+	public static async getErrorFromFetchResponse(error: Response | Error): Promise<RestClientCallError> {
+		if (error instanceof Error) {
+			const statusCode = 0;
+			const message = error.message || 'Unknown Error';
+			const statusMessage = (error['code'] || 'Unknown Error').toString();
+			return new RestClientCallError(message, statusCode, statusMessage, '');
+		} else {
+			const statusCode = error?.status || 0;
+			const statusMessage = (error?.statusText || 'Unknown Error').toString();
+			const body = (error.text && typeof error.text === 'function' && (await error.text())) || '';
```

Comment:
why do we need `await` here?

Reply:
https://developer.mozilla.org/en-US/docs/Web/API/Response/text
text() is an async method to pull the body